### PR TITLE
Layout Extraction and Modal.com

### PR DESCRIPTION
**Pull request by FFFiend**
Addresses #187 , coupled with Modal.com files

The file added as ```test.png``` is a screenshot taken of a publicly available, contract template from Google images.

### CODE REVIEW

**Comment by FFFiend** [comment]
https://huggingface.co/impira/layoutlm-document-qa 

This model is fine tuned for document question answering, and uses the same LayoutLM model to do so. It solves multiple problems such as conversion, data extraction etc but I'm testing it and it seems like its tuned to only give one word responses at max? See below:
```
nlp = pipeline("document-question-answering",
model = "impira/layoutlm-document-qa")
print(nlp("contract.png","Could you extract all key information from this document?"))

#output below 
[{'score': 0.0048963166773319244, 'answer': 'bills', 'start': 14, 'end': 14}]
```

**Comment by FFFiend** [comment]
As for parsing the output yourself, seems like we need to do training and fine tuning to fit the freight use case, after looking at a few examples online people use predefined labels and then use LayoutLM to extract predictions of those labels within the document, rather than extracting actual text from the output.

**Comment by FFFiend** [comment]
Here is the output you get after decoding the input ids for the Freight Contract document.
```
[CLS] hazardous material manifest truck : route truck 001 ( rt001 ) date : 9 / 25 / 2015 7 : 50 : 02 am manifest # : rt00120150925075002 departing location : driver : proper shipping name hazard pg rq quantity returned package special product code total weight class quantity permit ( lbs ) un proper shipping name hazard pg rq quantity returned package special product code total weight number class quantity pennit ( lbs ) sy ee, | es es ss | 5 a es [ 0 | _ | eee [ so es se ) ss ss es 2 | [SEP]
```

**Comment by FFFiend** [comment]
Screen Shot 2023-06-02 at 8 33 26 PM 

An image of how bounding boxes are shown on a load manifest document. It's not grabbing everything probably because the resolution of the file isn't small enough to be tokenized into 512 tokens. See below for a better result on a rate confirmation document.


Screen Shot 2023-06-02 at 8 35 05 PM

**Comment by FFFiend** [comment]
As for classification, this has to be done manually. Labels and colors must be initialized and then a correspondence has to be made between the boxes, input_ids and the labels. For example: {TABLE:5, CONTACT_INFO:3. ..}  where we have a frequency dict of each type of structure. Will look into this now

**Comment by FFFiend** [comment]
Screen Shot 2023-06-03 at 4 01 13 AM

Model mislabels mostly everything, as it encodes all the classified texts as 0's and 1's (which map to 'TABLE' and 'CONTACT_INFO' in my label_dict. Need to train on a freight dataset to get optimal results.

File: openadapt/strategies/mixins/freight_documents_test/Screenshot Dataset.v1i.coco/README.dataset.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Screenshot Dataset > 2023-06-03 1:20pm
```
Comment (abrichr) [inline openadapt/strategies/mixins/freight_documents_test/Screenshot Dataset.v1i.coco/README.dataset.txt:1]: Can you please move this (and other similar files) into the `resources` directory?

**Comment by FFFiend** [comment]
Currently experiencing an issue in regards to bounding boxes and words after processing data. Will prioritize today.

**Comment by FFFiend** [comment]
> @FFFiend can you please prioritize merging this?
> 
> Also why are the aspect ratios off on the images (leading to them looking "squished")?

I think they were stored in the dataset on roboflow this way.

**Comment by FFFiend** [comment]
Okay, so after coming back to this issue I realized that the above picture with just 1 type of label being classified was because I wasn't inferring properly. Here is a better result on an Invoice bill (in terms of classification):
Screen Shot 2023-06-17 at 4 24 59 AM

Results are not great... but it's a slight improvement from last time.

**Comment by FFFiend** [comment]
Question I asked GPT-4 on LayoutLMv2's off the shelf labels. 
```
The LayoutLMv2 model is equipped with off-the-shelf labels that are based on the FUNSD dataset, which can be generalized for form-like tasks, such as information extraction from documents, receipts, etc. The dataset has the following 4 main labels:

1. "S": for the "selected" elements of forms, representing filled-in information.
2. "B": for "beginning", representing the starting token or element of a labeled field in a form.
3. "I": for "inside", representing the intermediate tokens or elements within a labeled field in a form.
4. "O": for "outside", representing tokens or elements that do not belong to any labeled field in a form.

These labels are a variant of the standard BILOU (Beginning, Inside, Last, Outside, Unit-length) tagging scheme – typically required for labeling document entities using the LayoutLMv2 model. Note that LayoutLMv2 adapts to a specific task when fine-tuned on annotated data related to that task.
```

Note that for GUI, this model is unsuitable because it focuses specifically on text classification and extraction. I think going more towards Object Detection is what we want. So current action items would be:
```
1) Fine tune LayoutLMv2 on label format and form dataset (lots available on paperswithcode) similar to what we have in the commit currently.

2) YOLO/ CNN for Object Detection including training + inference on labelled GUI data.
```
The above only takes care of labels, and not **tables**. There's models trained solely for table classification such as https://huggingface.co/microsoft/table-transformer-detection. This can be simply incorporated by passing the image once into LayoutLMv2 and then into a table transformer to extract both.

**Comment by FFFiend** [comment]
Here's a couple HuggingFace spaces with different detection/extraction models:
https://huggingface.co/spaces/ShilongLiu/Grounding_DINO_demo

https://huggingface.co/spaces/nielsr/LayoutLMv2-FUNSD

**Comment by FFFiend** [comment]
I think this might be what we need for GUI object detection..seems easier to setup than gisting :D but their demo isn't live 

https://github.com/chenjshnn/Object-Detection-for-Graphical-User-Interface

**Comment by abrichr** [comment]
> I think this might be what we need for GUI object detection..seems easier to setup than gisting :D but their demo isn't live

I explored this about a year ago and didn't find much of value, author did not respond to emails. If you have any specific suggestions about how we can use it please clarify 🙏

**Comment by abrichr** [comment]
> https://huggingface.co/spaces/ShilongLiu/Grounding_DINO_demo

Great find! This looks very promising. Added in https://github.com/MLDSAI/OpenAdapt/issues/288

**Comment by abrichr** [comment]
> https://github.com/MLDSAI/OpenAdapt/issues/288

Can you please clarify how this is different from what you've already implemented?

**Comment by FFFiend** [comment]
> > https://huggingface.co/spaces/nielsr/LayoutLMv2-FUNSD
> 
> Can you please clarify how this is different from what you've already implemented?

No difference, just a convenient demo hosted on the web to try out instead of having to run the script each time. Sorry i should've elaborated in the original comment.



> > I think this might be what we need for GUI object detection..seems easier to setup than gisting :D but their demo isn't live
> 
> I explored this about a year ago and didn't find much of value, author did not respond to emails. If you have any specific suggestions about how we can use it please clarify 🙏

There's a repo linked in the Readme that contains the model along with a script to test it but I think it needs some refactoring and package updates.

**Comment by FFFiend** [comment]
Sure thing: Current action items are
```
1) Test and display output from Document Question-Answering Pipeline on the LayoutLMv3 model.
2) Train raw extraction on the LayoutLMv3 model.
```

**Comment by FFFiend** [comment]
https://github.com/huggingface/diffusers/issues/3828 Currently facing some issues with using pipelines on LayoutLMv3.

**Comment by FFFiend** [comment]
In the meantime, would you like to have a mixin that takes in an image along with a question (for document classification) using the original layout-lm model that seemed to be working well in the first reply to this PR?

**Comment by abrichr** [comment]
@FFFiend can you please follow up on the issue you posted to huggingface/diffusers?

Also in the future please include the full exception trace, not just the error message.

> In the meantime, would you like to have a mixin that takes in an image along with a question (for document classification) using the original layout-lm model that seemed to be working well in the first reply to this PR?

Let's try to get v3 working 🙏

**Comment by abrichr** [comment]
Closed in favor of https://github.com/OpenAdaptAI/OpenAdapt/pull/341

**Pull request closed without merging**